### PR TITLE
EOS-17950: Mini Provisioner: s3_setup Test phase enhancement

### PR DESCRIPTION
**Pull request by pratyush-seagate**
commit 3b3e12ef8838f2e81af2c7a0dd0ab827f0c5024d
Author: pratyush-seagate <pratyush.k.khan@seagate.com>
Date:   Thu Feb 25 07:34:26 2021 -0700

        EOS-17950: Mini Provisioner: s3_setup Test phase enhancement
            Change description:
                    Changes as described in ticket EOS-17950 to place the s3-sanity tests in python version of _s3_setup.
            Added/modified/deleted files:
                modified:   installhelper.sh
                deleted:    s3-sanity-test.sh
                modified:   scripts/provisioning/s3_setup
                modified:   scripts/provisioning/testcmd.py
                modified:   scripts/s3-sanity/s3-sanity-test.sh
        Unit test:
                Individual unit tests of scripts/s3-sanity/s3-sanity-test.sh with various combinations of input params.
                Run "s3_setup config" with these changes and verified steps worked. Performed complete mini-provisioner s3 steps.

    Signed-off-by: pratyush-seagate <pratyush.k.khan@seagate.com>

### CODE REVIEW

**Comment by pratyush-seagate** [comment]
Unit test output:

# ./s3-sanity-test.sh -h
USAGE: bash s3-sanity-test.sh
Run S3 sanity test.

PRE-REQUISITES: s3cmd and s3iamcli need to be present

where:
    -h     display this help and exit
    -c     clean s3 resources if present
    -e     specify the s3 server endpoint, pass 127.0.0.1 if running on s3 server node
           else provide data interface ip which needs to be in /etc/hosts
    -p     specify ldap password in text form, mandatory

Operations performed:
  * Create Account
  * Create User
  * Create Bucket
  * Put Object
  * Delete Object
  * Delete Bucket
  * Delete User
  * Delete Account
  * ldapsearch s3-background-delete-svc account



# ./s3-sanity-test.sh -p ldapadmin

S3IAMCLI is not found, both S3IAMCLI and S3CMD tests are disabled.
Both s3iamcli and s3cmd tests are disabled, nothing to config.
Both s3iamcli and s3cmd tests are disabled, nothing to cleanup.


*** S3 Sanity ***


***** S3: SANITY ldapsearch TESTS SUCCESSFULLY COMPLETED *****

Both s3iamcli and s3cmd tests are disabled, nothing to restore.


***** S3: SANITY ALL TESTS SUCCESSFULLY COMPLETED *****




# ./s3-sanity-test.sh -p ldapadmin

Check s3iamcli...OK, S3IAMCLI tests are enabled.

S3CMD is not found, S3CMD (data path) tests are disabled.

Sanity Account doesn't exist for cleanup


*** S3 Sanity ***


**** Create Account *******
UserId = AIDA7CCD8942D1B24144A, ARN = arn:aws:iam::210575358890:user/SanityUserToDeleteAfterUse, Path = /

Delete User - 'SanityUserToDeleteAfterUse':
User deleted.

Delete Account - 'SanityAccountToDeleteAfterUse':
Account deleted successfully.


***** S3: SANITY IAMCLI TESTS SUCCESSFULLY COMPLETED *****



***** S3: SANITY ldapsearch TESTS SUCCESSFULLY COMPLETED *****

/root/.sgs3iamcli/config.yaml


# ./s3-sanity-test.sh -p ldapadmin

Check s3iamcli...OK, S3IAMCLI tests are enabled.

Check S3CMD...OK, S3CMD data path tests are enabled.

Sanity Account doesn't exist for cleanup


*** S3 Sanity ***


**** Create Account *******
UserId = AIDA1522AD8DCAF0494AA, ARN = arn:aws:iam::718266320394:user/SanityUserToDeleteAfterUse, Path = /

Create bucket - 'sanitybucket':
ERROR: Error parsing xml: Malformed error XML returned from remote server..  ErrorXML: <html><body><h1>503 Service Unavailable</h1>
No server is available to handle this request.
</body></html>

WARNING: Retrying failed request: /?location (503 (Service Unavailable))
WARNING: Waiting 3 sec...
Bucket 's3://sanitybucket/' created
1+0 records in
1+0 records out
5000000 bytes (5.0 MB) copied, 0.0388631 s, 129 MB/s

Upload '/tmp/SanityObjectToDeleteAfterUse.input62476' to 'sanitybucket':
INFO: No cache file found, creating it.
INFO: Compiling list of local files...
INFO: Running stat() and reading/calculating MD5 values on 1 files, this may take some time...
INFO: Summary: 1 local files to upload
upload: '/tmp/SanityObjectToDeleteAfterUse.input62476' -> 's3://sanitybucket/SanityObjectToDeleteAfterUse'  [1 of 1]
 5000000 of 5000000   100% in    0s    34.78 MB/s  done

List uploaded SanityObjectToDeleteAfterUse in 'sanitybucket':
2021-02-25 14:06      5000000  s3://sanitybucket/SanityObjectToDeleteAfterUse

Download 'SanityObjectToDeleteAfterUse' from 'sanitybucket':
INFO: Summary: 1 remote files to download
download: 's3://sanitybucket/SanityObjectToDeleteAfterUse' -> '/tmp/SanityObjectToDeleteAfterUse.out62476'  [1 of 1]
 5000000 of 5000000   100% in    0s    80.96 MB/s  done

Data integrity check: Passed.

Delete 'SanityObjectToDeleteAfterUse' from 'sanitybucket':
INFO: Summary: 1 remote files to delete
delete: 's3://sanitybucket/SanityObjectToDeleteAfterUse'

Delete bucket - 'sanitybucket':
Bucket 's3://sanitybucket/' removed


***** S3: SANITY DATA TESTS SUCCESSFULLY COMPLETED *****


Delete User - 'SanityUserToDeleteAfterUse':
User deleted.

Delete Account - 'SanityAccountToDeleteAfterUse':
Account deleted successfully.


***** S3: SANITY IAMCLI TESTS SUCCESSFULLY COMPLETED *****



***** S3: SANITY ldapsearch TESTS SUCCESSFULLY COMPLETED *****

/root/.sgs3iamcli/config.yaml
/tmp/.s3cfg
/tmp/SanityObjectToDeleteAfterUse.input62476
/tmp/SanityObjectToDeleteAfterUse.out62476


***** S3: SANITY ALL TESTS SUCCESSFULLY COMPLETED *****


# ./s3-sanity-test.sh -p ldapadmin -e 127.0.0.1

Check s3iamcli...OK, S3IAMCLI tests are enabled.

Check S3CMD...OK, S3CMD data path tests are enabled.
using s3endpoint 127.0.0.1
/root/.sgs3iamcli/config.yaml

Sanity Account doesn't exist for cleanup


*** S3 Sanity ***


**** Create Account *******
UserId = AIDA077D151EEBE44736B, ARN = arn:aws:iam::528228410251:user/SanityUserToDeleteAfterUse, Path = /

Create bucket - 'sanitybucket':
Bucket 's3://sanitybucket/' created
1+0 records in
1+0 records out
5000000 bytes (5.0 MB) copied, 0.0436143 s, 115 MB/s

Upload '/tmp/SanityObjectToDeleteAfterUse.input65088' to 'sanitybucket':
INFO: No cache file found, creating it.
INFO: Compiling list of local files...
INFO: Running stat() and reading/calculating MD5 values on 1 files, this may take some time...
INFO: Summary: 1 local files to upload
upload: '/tmp/SanityObjectToDeleteAfterUse.input65088' -> 's3://sanitybucket/SanityObjectToDeleteAfterUse'  [1 of 1]
 5000000 of 5000000   100% in    0s    34.34 MB/s  done

List uploaded SanityObjectToDeleteAfterUse in 'sanitybucket':
2021-02-25 14:08      5000000  s3://sanitybucket/SanityObjectToDeleteAfterUse

Download 'SanityObjectToDeleteAfterUse' from 'sanitybucket':
INFO: Summary: 1 remote files to download
download: 's3://sanitybucket/SanityObjectToDeleteAfterUse' -> '/tmp/SanityObjectToDeleteAfterUse.out65088'  [1 of 1]
 5000000 of 5000000   100% in    0s    71.25 MB/s  done

Data integrity check: Passed.

Delete 'SanityObjectToDeleteAfterUse' from 'sanitybucket':
INFO: Summary: 1 remote files to delete
delete: 's3://sanitybucket/SanityObjectToDeleteAfterUse'

Delete bucket - 'sanitybucket':
Bucket 's3://sanitybucket/' removed


***** S3: SANITY DATA TESTS SUCCESSFULLY COMPLETED *****


Delete User - 'SanityUserToDeleteAfterUse':
User deleted.

Delete Account - 'SanityAccountToDeleteAfterUse':
Account deleted successfully.


***** S3: SANITY IAMCLI TESTS SUCCESSFULLY COMPLETED *****



***** S3: SANITY ldapsearch TESTS SUCCESSFULLY COMPLETED *****

/root/.sgs3iamcli/config.yaml
/tmp/.s3cfg
/tmp/SanityObjectToDeleteAfterUse.input65088
/tmp/SanityObjectToDeleteAfterUse.out65088


***** S3: SANITY ALL TESTS SUCCESSFULLY COMPLETED *****

**Comment by pratyush-seagate** [comment]
Manually started premerge at http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 1]
```diff
@@ -17,81 +18,301 @@
...
+  * Delete Account
+  * ldapsearch s3-background-delete-svc account"
+
+die_with_error ()
+{
+  echo "$@" 1>&2
+  exit -1
+}
+
+update_config() {
```
Comment (nileshgovande) [inline scripts/s3-sanity/s3-sanity-test.sh:69]: Write comments - what this function is doing

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 2]
```diff
@@ -17,81 +18,301 @@
...
+    echo -e "\nCheck S3CMD...OK, S3CMD data path tests are enabled."
+    s3cmdtestson=true
+  else
+    echo -e "\nS3CMD is not found, S3CMD (data path) tests are disabled."
+  fi
+else
+  echo -e "\nS3IAMCLI is not found, both S3IAMCLI and S3CMD tests are disabled."
+fi
+
+systemctl restart s3authserver
```
Comment (nileshgovande) [inline scripts/s3-sanity/s3-sanity-test.sh:191]: Remove

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 3]
```diff
@@ -17,81 +18,301 @@
...
+      $TEST_CMD rb "s3://sanitybucket" >/dev/null 2>&1 || echo ""
+    fi
+
+    s3iamcli deleteuser -n SanityUserToDeleteAfterUse --access_key "$access_key" --secret_key "$secret_key" --no-ssl 2> /dev/null || echo ""
+    s3iamcli deleteaccount -n SanityAccountToDeleteAfterUse --access_key "$access_key" --secret_key "$secret_key" --no-ssl 2> /dev/null || echo ""
+  fi
+
+  restore_config
+}
+
```
Comment (nileshgovande) [inline scripts/s3-sanity/s3-sanity-test.sh:148]: Mention this as entry point

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 4]
```diff
@@ -17,81 +18,301 @@
...
+    fi
 
-# create a test file
-test_file_input=/tmp/"$object_name".input
-test_output_file=/tmp/"$object_name".out
+    content_md5_after=$(md5sum $test_output_file | cut -d ' ' -f 1)
+    echo -en "\nData integrity check: "
+    [[ $content_md5_before == $content_md5_after ]] && echo 'Passed.'
+    echo -e "\nDelete 'SanityObjectToDeleteAfterUse' from 'sanitybucket': "
+    $TEST_CMD del "s3://sanitybucket/SanityObjectToDeleteAfterUse"
```
Comment (nileshgovande) [inline scripts/s3-sanity/s3-sanity-test.sh:259]: Make sure object is completely deleted - either using curl / REST / STs - I leave it to you

Comment (nileshgovande) [reply scripts/s3-sanity/s3-sanity-test.sh:259]: bgdel verification not done (Write TODO)  - Create a seperate JIRA

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 5]
```diff
@@ -17,81 +18,301 @@
...
-#!/bin/bash
-
-set -e
 
 #########################
 # S3 sanity test script #
 #########################
 
-USAGE="USAGE: bash $(basename "$0") [--help]
+ldappasswd=''
```
Comment (nileshgovande) [inline scripts/s3-sanity/s3-sanity-test.sh:26]: Restructure the logic for s3iamclitestson s3cmdtestson

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 6]
```diff
@@ -1,3 +1,4 @@
+#!/bin/bash
```
Comment (nileshgovande) [inline scripts/s3-sanity/s3-sanity-test.sh:1]: if(s3imacli)
{
    create acc
    create user
      if(s3cmd)
      {
      create bucket
      create object
      delete bucket
      delete object
      }
    delete accd
    delete user  
}

Comment (pratyush-seagate) [reply scripts/s3-sanity/s3-sanity-test.sh:1]: The previous code was following this pattern. Please review the current code which is slightly restructured.

File: scripts/s3-sanity/s3-sanity-test.sh (hunk 5 above)
Comment (pratyush-seagate) [reply scripts/s3-sanity/s3-sanity-test.sh:26]: Please review the current code which is slightly restructured.

File: scripts/s3-sanity/s3-sanity-test.sh (hunk 4 above)
Comment (pratyush-seagate) [reply scripts/s3-sanity/s3-sanity-test.sh]: created https://jts.seagate.com/browse/EOS-18464

**Review by AjinkyaDhumal — APPROVED** [review]

**Review by nileshgovande — APPROVED** [review]

**Pull request merged**